### ISSUE TITLE
[Feature Request] running average backward()

### ISSUE THREAD

**Issue by dem123456789** [pytorch/pytorch#15097]
## 🚀 Feature
Accumulated the gradient of model parameters by weighted average `.grad` and new gradients with respect to the number of samples.

## Motivation
We may want to do `loss.backward()` multiple times before we update model parameters by `optimizer.step()` and `optimizer.zero_grad()` due to using small batch size.
## Pitch
`loss.backward()` and `optimizer.zero_grad()` do not need to synchronize like in Truncated BPTT or using small batch size due to memory consumption.
We can have `loss.backward(batch_size=128)` or `optimizer.backward(loss, batch_size=128)`. Then we maintain a counter for the number of samples used to compute gradients. Then we can perform running average on `.grad` when we call `.backward()` the second time.

## Additional context
I think it would be possible the statistics of gradient become more and more important. Because many theories in statistcs and optimization are based on the statistics of the first order and even second order gradients. For example, we can compute Fisher Information by using the hessian or the covariance of the first order gradients. So far we only record batch mean of gradients, it may be possible to extend to more complex statistics and even per-sample, per-parameter high-order gradients in the future.

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7

**Comment by SsnL** [comment]
Can't you just do this before calling optimizer?

```py
for p in model.parameters():
  if p.grad is not None:
    p.grad.div_(niter_since_last_optim_step)
```

**Comment by dem123456789** [comment]
This is assuming the number of samples that used to compute gradients are always the same. If I have a batch of data that have less data that I use before, this won't work.

**Comment by albanD** [comment]
Note that an equivalent way to do `loss.backward(batch_size=128)` is simply to do `(loss / 128).backward()` which should be simple enough to do.

Feel free to re-open if this is not possible for your use case and you need more help!
